mdformat/renderer/__init__.py

```diff
@@ -99,8 +105,27 @@ def render(
         rendered_content = text_stack.pop()
         assert not text_stack, "Text stack should be empty before returning"
 
-        if not _recursion_level:
+        if finalize and not _recursion_level:
             rendered_content = removesuffix(rendered_content, MARKERS.BLOCK_SEPARATOR)
             rendered_content = rendered_content.replace(MARKERS.BLOCK_SEPARATOR, "\n\n")
+
+            if options.get("mdformat", {}).get("keep_references", False) and env.get(
+                "references", None
+            ):
+                rendered_content += "\n\n"
+                rendered_content += self.write_references(env)
+
             rendered_content += "\n"
+
         return rendered_content
+
+    @staticmethod
+    def write_references(env: dict) -> str:
```

Comment:
Does this need to be exported? I'd prefer it to be private method if not.

Reply:
Why not allow it to be overridden in a subclass?
In fact, technically I'd say it's a bit of a shortcoming of this renderer; that you can't override any of the builtin token render methods in a subclass, as opposed to https://github.com/executablebooks/markdown-it-py/blob/master/markdown_it/renderer.py where the render method for every token can be overridden.
There's not an immediate use case for that though, and you can now achieve it via plugins, so I'm not suggesting changing it, just noting 😉

Reply:
So by default I always want to make everything private to avoid a maintenance nightmare. Any changes, refactoring etc. that change public API is a breaking change and forces me to make a major version bump, and is a maintenance hassle for a potential user.

> In fact, technically I'd say it's a bit of a shortcoming of this renderer;

It is a shortcoming for someone willing to subclass the class. It is also a major advantage for me trying to maintain the project. So far someone subclassing the renderer class is hypothetical, but me maintaining it is reality.

I didn't really intend to make even `MDRenderer.render` public API, it just happened to be useful for the plugin interface. I mostly care about having the formatter API public.

Reply:
I'm not fully opposed to making any API public, I just think there should usually be a good use-case and hopefully also at least one user willing to use it.

Reply:
sure, its not a hill I want to die on lol; changed in 895d3106338c543d4267f5976af7978a3fdecaa6

Reply:
:smile: nice thanks!